### PR TITLE
feat(FQDN): Implement of struct host_port_group

### PR DESCRIPTION
**Pull request by GehaFearless**
issue: https://github.com/apache/incubator-pegasus/issues/1404

Implement of struct `group_host_port` used by 'rpc_host_port'.
Also add some code on class `rpc_host_port` whose be related to `group_host_port`
used.

### CODE REVIEW

File: src/runtime/rpc/group_host_port.h [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+private:
+    typedef std::vector<host_port> members_t;
+    typedef utils::auto_read_lock alr_t;
+    typedef utils::auto_write_lock alw_t;
+
+    mutable utils::rw_lock_nr _lock;
+    members_t _members;
+    int _leader_index;
+    bool _update_leader_automatically;
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:85]: Add some necessary comments to these variables, they are not very self-evident.

File: src/runtime/rpc/group_host_port.h [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                                : _members[rand::next_u32(0, (uint32_t)_members.size() - 1)];
+    }
+    host_port next(host_port current) const;
+    host_port leader() const
+    {
+        alr_t l(_lock);
+        return _leader_index >= 0 ? _members[_leader_index] : host_port::s_invalid_host_port;
+    }
+    void leader_forward();
+    host_port possible_leader();
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:72]: Add some necessary comments to these functions, they are not very self-evident.

File: src/runtime/rpc/group_host_port.h [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+inline void rpc_group_host_port::set_leader(host_port hp)
+{
+    alw_t l(_lock);
+    if (hp.is_invalid()) {
+        _leader_index = -1;
+        return;
+    }
+
+    CHECK_EQ_MSG(hp.type(), HOST_TYPE_IPV4, "rpc group host_port member must be ipv4");
+    for (int i = 0; i < (int)_members.size(); i++) {
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:149]:
```suggestion
    for (int i = 0; i < static_cast<int>(_members.size()); i++) {
```
Some other places are the same.

File: src/runtime/rpc/group_host_port.h [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+inline bool rpc_group_host_port::remove(host_port hp)
+{
+    alw_t l(_lock);
+    auto it = std::find(_members.begin(), _members.end(), hp);
+    if (it == _members.end()) {
+        return false;
+    }
+
+    if (-1 != _leader_index && hp == _members[_leader_index]) {
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:180]: How about define `-1` as a static const variable like `kInvalidIndex`?

File: src/runtime/rpc/rpc_host_port.cpp [hunk 5]
```diff
@@ -43,8 +44,17 @@ host_port::host_port(rpc_address addr)
               addr.ipv4_str());
         _port = addr.port();
     } break;
-    case HOST_TYPE_GROUP:
-        CHECK(false, "type HOST_TYPE_GROUP not support!");
+    case HOST_TYPE_GROUP: {
+        auto group_address = addr.group_address();
+        *this = host_port();
```
Comment (acelyc111) [inline src/runtime/rpc/rpc_host_port.cpp:49]: This line can be removed?

File: src/runtime/rpc/rpc_host_port.cpp [hunk 6]
```diff
@@ -43,8 +44,17 @@ host_port::host_port(rpc_address addr)
...
+        this->assign_group(group_address->name());
```
Comment (acelyc111) [inline src/runtime/rpc/rpc_host_port.cpp:50]: `this->` can be omitted?

File: src/runtime/rpc/rpc_host_port.cpp [hunk 7]
```diff
@@ -43,8 +44,17 @@ host_port::host_port(rpc_address addr)
...
+        for (const auto &address : group_address->members()) {
+            CHECK_TRUE(this->group_host_port()->add(host_port(address)));
+        }
+        this->group_host_port()->set_update_leader_automatically(
+            group_address->is_update_leader_automatically());
+        this->group_host_port()->set_leader(host_port(group_address->leader()));
```
Comment (acelyc111) [inline src/runtime/rpc/rpc_host_port.cpp:56]: How about encapsulate these code as the c'stor of group_host_port?

File: src/runtime/rpc/rpc_host_port.cpp [hunk 8]
```diff
@@ -79,7 +90,9 @@ host_port &host_port::operator=(const host_port &other)
         _port = other.port();
         break;
     case HOST_TYPE_GROUP:
-        CHECK(false, "type HOST_TYPE_GROUP not support!");
+        _group_host_port = other.group_host_port();
+        group_host_port()->add_ref();
```
Comment (acelyc111) [inline src/runtime/rpc/rpc_host_port.cpp:94]: How about add_ref() in the equal operator of group_host_port?

File: src/runtime/rpc/group_host_port.h [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        for (const auto &hp : hps) {
+            LOG_WARNING_IF(!add(hp), "duplicate adress {}", hp);
+        }
+    }
+    void set_leader(host_port hp);
+    bool remove(host_port hp) WARN_UNUSED_RESULT;
+    bool contains(host_port hp) const WARN_UNUSED_RESULT;
+    int count() const;
+
+    const std::vector<host_port> &members() const { return _members; }
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:66]:
```suggestion
    const std::vector<host_port> &members() const
    {
        alr_t l(_lock);
        return _members;
    }
```

File: src/runtime/rpc/group_host_port.h [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+        return host_port::s_invalid_host_port;
+    }
+
+    if (current.is_invalid()) {
+        return _members[rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1))];
+    }
+
+    auto it = std::find(_members.begin(), _members.end(), current);
+    if (it == _members.end()) {
+        return _members[rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1))];
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:237]: `rand::next_u32(0, static_cast<uint32_t>(_members.size() - 1))` apears 4 times, let's define it as a function like `uint32_t random_index_unlocked()`?

File: src/runtime/rpc/group_host_port.h [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    const char *name() const { return _name.c_str(); }
+
+private:
+    typedef std::vector<host_port> members_t;
+    typedef utils::auto_read_lock alr_t;
+    typedef utils::auto_write_lock alw_t;
+
+    mutable utils::rw_lock_nr _lock;
+    members_t _members;
+    int _leader_index;
```
Comment (acelyc111) [inline src/runtime/rpc/group_host_port.h:98]: Add some comments for `_leader_index`, for example mention that it's not always valid even if `_members` is not empty.

**Review by acelyc111 — DISMISSED** [review]

File: src/runtime/rpc/rpc_host_port.h [hunk 12]
```diff
@@ -57,10 +59,18 @@ class host_port
...
+        CHECK_NOTNULL(_group_host_port, "group_host_port cannot be null!");
+        return _group_host_port;
+    }
+    void assign_group(const char *name);
+
 private:
     std::string _host;
     uint16_t _port = 0;
     dsn_host_type_t _type = HOST_TYPE_INVALID;
+    rpc_group_host_port *_group_host_port = nullptr;
```
Comment (empiredan) [inline src/runtime/rpc/rpc_host_port.h:73]: Use ref_ptr<rpc_group_host_port> to manage `ref_counter`-typed pointer ? Or just shared_ptr ?

File: src/runtime/rpc/group_host_port.h [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+    // failure_detector should avoid failure detecting logic is affected by rpc failure or rpc
+    // forwarding. So we need a switch to contronl update leader automatically.
+    bool is_update_leader_automatically() const { return _update_leader_automatically; }
+    void set_update_leader_automatically(bool value) { _update_leader_automatically = value; }
+    const char *name() const { return _name.c_str(); }
+
+private:
+    typedef std::vector<host_port> members_t;
+    typedef utils::auto_read_lock alr_t;
```
[the remark this answers is not in the document]
Comment (GehaFearless) [reply src/runtime/rpc/group_host_port.h:98]: Be like `src/runtime/rpc/group_address.h`，could I change it？

**Review by acelyc111 — APPROVED** [review]

**Review by empiredan — APPROVED** [review]

**Pull request merged**